### PR TITLE
MKLDNN Forward FullyConnected  op cache

### PR DESCRIPTION
**Pull request by huangzhiyuan**
## Description ##
Enable primitive allocation cache for FullyConnected, separate from PR [#11301](https://github.com/apache/incubator-mxnet/pull/11301#discussion_r200819757) 

@zheng-da , @pengzhao-intel, @ZhennanQin

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by marcoabreu** [comment]
Hi, please add a test to ensure the caching works as expected

**Comment by BiranLi** [comment]
Which version of mkldnn is suitable for this PR?
Is the 1.4 version working?

**Comment by huangzhiyuan** [comment]
The PR is based on release v0.14 version. Thanks.

**Comment by eric-haibin-lin** [comment]
@azai91 could you help review

File: src/operator/nn/mkldnn/mkldnn_fully_connected.cc [hunk 1]
```diff
@@ -82,6 +82,101 @@ inline static mkldnn::inner_product_backward_weights::primitive_desc GetIPBwdWei
...
+    else
+      this->out->set_data_handle(output.get_data_handle());
+
+    if (bias != nullptr) {
+      if (this->bias == nullptr)
+        this->bias = std::shared_ptr<mkldnn::memory>(new mkldnn::memory(
+            ipFwd_pd.bias_primitive_desc(), bias->get_data_handle()));
+      else
+        this->bias->set_data_handle(bias->get_data_handle());
+      if (this->ipFwd == nullptr)
```
Comment (azai91) [inline src/operator/nn/mkldnn/mkldnn_fully_connected.cc]: fix indent here?

File: src/operator/nn/mkldnn/mkldnn_fully_connected.cc [hunk 2]
```diff
@@ -112,21 +207,21 @@ void MKLDNNFCForward(const nnvm::NodeAttrs& attrs, const OpContext &ctx,
...
+      GetFCFwd(attrs, data, weight, param.no_bias ? nullptr : &in_data[fullc::kBias],
+               out_md, ctx.is_train);
+  auto data_mem = data.GetMKLDNNDataReorder(FCFwd.ipFwd_pd.src_primitive_desc());
+  auto weight_mem = weight.GetMKLDNNDataReorder(FCFwd.ipFwd_pd.weights_primitive_desc());
   auto out_mem = CreateMKLDNNMem(out_data[fullc::kOut],
-      ipFwd_pd.dst_primitive_desc(), req[fullc::kOut]);
-  if (param.no_bias) {
-    MKLDNNStream::Get()->RegisterPrim(mkldnn::inner_product_forward(
-          ipFwd_pd, *data_mem, *weight_mem, *out_mem.second));
+      FCFwd.ipFwd_pd.dst_primitive_desc(), req[fullc::kOut]);
```
Comment (azai91) [inline src/operator/nn/mkldnn/mkldnn_fully_connected.cc]: pass in_data as last argument for CreateMKLDNNMem in case req is WriteInPlace

**Comment by azai91** [comment]
approve conditional on adding that param to CreateMKLDNNMem so we can support kWriteInPlace.

File: src/operator/nn/mkldnn/mkldnn_fully_connected.cc (hunk 1 above)
Comment (huangzhiyuan) [reply src/operator/nn/mkldnn/mkldnn_fully_connected.cc]: Have fixed.

File: src/operator/nn/mkldnn/mkldnn_fully_connected.cc (hunk 2 above)
Comment (huangzhiyuan) [reply src/operator/nn/mkldnn/mkldnn_fully_connected.cc]: Done.

**Pull request merged**